.github/workflows/ci.yml

```diff
@@ -351,3 +362,24 @@ jobs:
               | tee -a $GITHUB_STEP_SUMMARY >&2
             exit 1
           fi
+
+  generate_cache:
```

Comment:
Why do we need this separate job? Will the caching not happen automatically the first time a job/step is run that uses `cargo metadata` (ie, the "Set toolchain version" step)?

Reply:
Because Git actions start the jobs simultaneously and the cache is written back after the build and test processes, the jobs can't utilize the cache until it's been written back, potentially leading to network timeouts. To address this, I've modified the job `build and test` while also waiting for the completion of the `cargo metadata` operation and the successful writing of the cache to GitHub. This adjustment generates the cache before the jobs start, allowing them to retrieve it and thus avoiding any potential network timeout issues.

Reply:
Ah I see. I was assuming that the cache would be kept around between PR job runs, so while the first run of jobs might still have problems, subsequent runs would be fine. It looks like that [might be how it works](https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows), but now that I'm thinking about it more, it strikes me as cleaner to have a separate "generate cache" step even if it's technically unnecessary.